### DISCUSSION TITLE
Consolidate query editors

### DISCUSSION THREAD

**Discussion by marcusolsson** [grafana/grafana#38140] [Dashboards, Panels & Visualizations]
**What would you like to be added**:

When building a data source plugin, the QueryEditor should be used for Annotations and Query Variables by default.

This is already done for Explore, where the QueryEditor is used for Explore unless the plugin author explicitly overrides it.

**Why is this needed**:

To support annotations and query variables in a data source plugin, you'd have to create separate editors, which results in duplicated code. Using the QueryEditor by default would result in higher quality data sources that would be easier to maintain.

The only thing the plugin author would have to do to support annotations and variables is to implement

```ts
async metricFindQuery(query: MyQuery, options?: any): Promise<MetricFindValue[]>
```

and 

```ts
async annotationQuery(options: AnnotationQueryRequest<MyQuery>): Promise<AnnotationEvent[]>
```

Both would accept the query model, `MyQuery`, as defined by the `QueryEditor`.

**Comment by stevenvachon** [comment]
Maybe even a generated UI from a JSON/etc config file as well. I'd be interested to work on this if no one gets around to it by... the new year?

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
